apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+    view = findViewFromName(view.partOf);
+  }
+
+  debug('findMainView() => %s', view.name);
+  return view;
+}
 
-function nextQueuedPanel() {
-  if (queuedPanel) {
-    var next = queuedPanel;
-    queuedPanel = null;
-    Navigation.toPanel(next.panel, next.args)
-      .then(next.defer.resolve, next.defer.reject);
+/**
+ * @type {Defer}
+ * Used at startup time to indicate when the app loaded all dependencies.
+ * Will be nullified once the app is ready.
+ * @see Navigation.setReady
+ * @see waitForReady
+ */
+var readyDefer = Utils.Promise.defer();
+
+/**
+ * @returns {Promise} Resolved when the app is ready, rejected if the app is
+ * ready at call time.
+ */
+function waitForReady() {
+  if (!readyDefer) {
+    return Promise.reject(
+      new Error('waitForReady was called but status is ready already.')
+    );
   }
+
+  return readyDefer.promise;
 }
 
-var Navigation = {
-  isReady: false,
-  panelObjects: window,
-
-  defaultPanel: 'thread-list',
-
-  panels: {
-    'thread': {
-      behaviour: 'ConversationView',
-      wrapperPosition: 'left',
-      container: 'thread-messages'
-    },
-    'thread-list': {
-      behaviour: 'InboxView',
-      wrapperPosition: 'right',
-      container: 'thread-list'
-    },
-    'composer': {
-      behaviour: 'ConversationView',
-      wrapperPosition: 'left',
-      container: 'thread-messages'
-    },
-    'group-view': {
-      behaviour: 'GroupView',
-      wrapperPosition: 'left',
-      container: 'information-participants'
-    },
-    'report-view': {
-      behaviour: 'ReportView',
-      wrapperPosition: 'left',
-      container: 'information-report'
+/**
+ * Sets up the state to handle the navigation.
+ * Generally we're waiting for any ongoing transition to end before starting a
+ * new one. However if the ongoing transition was started with opts.partial ==
+ * true, we merely augment the ongoing transition with the new information,
+ * especially arguments.
+ *
+ * @param {Object} opts options
+ * @param {String} [opts.newPanel] The panel we'll switch to. This can be
+ * omitted in case we only want to execute beforeLeave.
+ * @param {Object} [opts.args] The arguments we'll use for the next panel.
+ * @param {Boolean} [opts.partial] If present and true, this new navigation
+ * request is a partial one missing information (eg arguments) and we'll call
+ * startNavigation once again without this property once we'll know the missing
+ * bits.
+ *
+ * @returns {Promise} Resolved when we could start the navigation (when an
+ * previous ongoing one is finished).
+ */
+function startNavigation(opts) {
+  var { newPanel, args, partial } = opts;
+
+  args = args || {};
+
+  debug(
+    'startNavigation(newPanel=%s, partial=%s)',
+    newPanel && newPanel.name, partial
+  );
+  if (navigationTransition && (!navigationTransition.partial || partial)) {
+    debug('startNavigation: previous navigation exists, deferring.');
+    return navigationTransition.defer.promise.then(
```

Comment:
__question:__ repeating my old question here, to at least decide that we don't want to change anything here :)

> Mmm, should not we chain in the order here? I mean if we call startNavigation 3 times (navigation1, navigation2, navigation3) while navigation4 is happening, then once it's finished, we proceed with navigation1 and reschedule navigation2 and navigation3 once again or I missing something?